### PR TITLE
Added a Default listener

### PR DESCRIPTION
**Pull request by upperwal**
Solving #377 

Added IPV4 - TCP on a random port as a default listener (```/ip4/0.0.0.0/tcp/0```)

@Stebalien can you please review.

### CODE REVIEW

File: defaults.go [hunk 1]
```diff
@@ -52,6 +53,15 @@ var RandomIdentity = func(cfg *Config) error {
 	return cfg.Apply(Identity(priv))
 }
 
+// DefaultListenAddrs configures libp2p to use default listen address
+var DefaultListenAddrs = func(cfg *Config) error {
+	defaultListenAddr, err := multiaddr.NewMultiaddr("/ip4/0.0.0.0/tcp/0")
```
Comment (Stebalien) [inline defaults.go]: This should also support `ip6`.

File: defaults.go [hunk 2]
```diff
@@ -80,6 +90,10 @@ var defaults = []struct {
 		fallback: func(cfg *Config) bool { return cfg.Peerstore == nil },
 		opt:      DefaultPeerstore,
 	},
+	{
+		fallback: func(cfg *Config) bool { return cfg.ListenAddrs == nil },
```
Comment (Stebalien) [inline defaults.go]: We should probably only apply these if we are using the default transports. I'd put this *before* the default transport section and check to see if `Transports` is nil.

**Comment by upperwal** [comment]
@Stebalien done.

File: libp2p_test.go [hunk 3]
```diff
@@ -39,6 +40,39 @@ func TestInsecure(t *testing.T) {
 	h.Close()
 }
 
+func TestDefaultListenAddrs(t *testing.T) {
```
Comment (Stebalien) [inline libp2p_test.go]: Should have a "correctly set the default listeners" test (i.e., make sure this actually works).

Comment (upperwal) [reply libp2p_test.go]: Something like this?
```go
h, err := New(ctx)
if err != nil {
	t.Fatal(err)
}
if len(h.Addrs()) != 2 {
	t.Error("expected 2 default listen addrs")
}
h.Close()
```

Is length of ```h.Addrs()``` sufficient for tests or should they be compared with the actual strings?

**Comment by upperwal** [comment]
@Stebalien Is jenkins only responsible for tests w/o compiling?

File: libp2p_test.go (hunk 3 above)
Comment (Stebalien) [reply libp2p_test.go]: That's probably good enough. Thanks!

Comment (Stebalien) [reply libp2p_test.go]: I think that test error is due to how we expand ALL addresses using the interface addresses in `Host.Addrs()`. Try calling `h.Network().ListenAddresses()`. In that case, we should even be able to verify that they are the correct addresses.

**Comment by upperwal** [comment]
@Stebalien My assumption was incorrect. I was assuming that if ```New``` is executed with two listener addresses, something like

```go
host, err := libp2p.New(
	ctx,
	libp2p.ListenAddrStrings("ip4/0.0.0.0/tcp/0", "ip6/0.0.0.0/tcp/0"),
)
```

then ```len(host.Addrs())``` would be 2 where one would be for ```ip4``` interface and the other one for ```ip6``` interface but it turns out that because I don't have an ipv6 connection ```host.Addrs()``` corresponds to my ```ipv4``` interface with one being ```127.0.0.1``` and the other one ```192.168.x.x```. 

For ```ip6/0.0.0.0/tcp/0```, ```AddListenAddr``` returns ```no suitable address found``` because I don't have any ```ipv6``` connection.

So all my tests are incorrect. What do you suggest. How should I test?

Test could be designed in a way that if either one of ```/ip4/0.0.0.0/tcp/x``` or ```/ip6/0.0.0.0/tcp/x``` is found in ```host.Network().ListenAddresses()``` it would be a success. This could be done by searching for ```/ip4/0.0.0.0/tcp/``` and ```/ip6/0.0.0.0/tcp/``` string in ```host.Network().ListenAddresses()```. Do you agree?

**Comment by Stebalien** [comment]
I'd check for `/ip4/0.0.0.0/tcp/x` in `Network().ListenAddresses()` and leave it at that. Nobody disables IPv4.

File: defaults.go [hunk 4]
```diff
@@ -52,6 +53,23 @@ var RandomIdentity = func(cfg *Config) error {
...
 }
 
+// DefaultListenAddrs configures libp2p to use default listen address
+var DefaultListenAddrs = func(cfg *Config) error {
+	defaultIP4ListenAddr, err := multiaddr.NewMultiaddr("/ip4/0.0.0.0/tcp/0")
+	if err != nil {
+		return err
+	}
+
+	defaultIP6ListenAddr, err := multiaddr.NewMultiaddr("/ip6/0.0.0.0/tcp/0")
```
Comment (Stebalien) [inline defaults.go]: This should be `/ip6/::/tcp/0` (sorry for not noticing earlier). You'll also have to fix the test. `0.0.0.0` will use the IP4 over IP6 ALL address.

Comment (upperwal) [reply defaults.go]: aah, my bad.

**Pull request merged**